File: apps/backend-python/src/housing_platform/shared/rate_limit.py
from datetime import UTC, datetime

from sqlalchemy.dialects.postgresql import insert
from sqlalchemy.orm import Session

from housing_platform.auth.errors import RateLimitedError
from housing_platform.db.models import ApiRateLimits


class RateLimitService:
    def __init__(self, db: Session) -> None:
        self._db = db

    def assert_rate_limit(self, bucket: str, max_requests: int, window_seconds: int) -> None:
        trimmed = bucket.strip()
        if not trimmed or max_requests <= 0 or window_seconds <= 0:
            raise RateLimitedError("Rate limit exceeded")

        now = datetime.now(UTC)
        window_start_epoch = (now.timestamp() // window_seconds) * window_seconds
        window_start = datetime.fromtimestamp(window_start_epoch, UTC)

        stmt = (
            insert(ApiRateLimits)
            .values(bucket=trimmed, window_start=window_start, request_count=1)
            .on_conflict_do_update(
                index_elements=[ApiRateLimits.bucket, ApiRateLimits.window_start],
                set_={"request_count": ApiRateLimits.request_count + 1},
            )
            .returning(ApiRateLimits.request_count)
        )
        count = self._db.execute(stmt).scalar_one()
        if count > max_requests:
            raise RateLimitedError("Rate limit exceeded")

File: apps/backend-go/internal/shared/ratelimit.go
package shared

import (
	"context"
	"strings"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	apierrors "github.com/housing-platform/backend-go/internal/api/errors"
)

type RateLimitService struct {
	pool *pgxpool.Pool
}

func NewRateLimitService(pool *pgxpool.Pool) *RateLimitService {
	return &RateLimitService{pool: pool}
}

func (s *RateLimitService) AssertRateLimit(ctx context.Context, tx pgx.Tx, bucket string, maxRequests int, windowSeconds int) error {
	trimmed := strings.TrimSpace(bucket)
	if trimmed == "" || maxRequests <= 0 || windowSeconds <= 0 {
		panic(apierrors.RateLimited("Rate limit exceeded"))
	}

	now := time.Now().UTC()
	windowStartEpoch := (now.Unix() / int64(windowSeconds)) * int64(windowSeconds)
	windowStart := time.Unix(windowStartEpoch, 0).UTC()

	var count int32
	err := tx.QueryRow(ctx, `
		INSERT INTO api_rate_limits (bucket, window_start, request_count)
		VALUES ($1, $2, 1)
		ON CONFLICT (bucket, window_start)
		DO UPDATE SET request_count = api_rate_limits.request_count + 1
		RETURNING request_count
	`, trimmed, windowStart).Scan(&count)
	if err != nil {
		return err
	}

	if count > int32(maxRequests) {
		panic(apierrors.RateLimited("Rate limit exceeded"))
	}

	return nil
}
